### PR TITLE
Attach res

### PR DESCRIPTION
**Pull request by yerbol-akhmetov**
## Changes proposed in this Pull Request
Here, I propose a modification to `attach_wind_and_solar` function in `add_electricity.py`. With this modification, it would be possible to set capacities for renewables from `custom_powerplants.csv`. Currently, the only possible option to set the renewable capacities is using `estimate_renewable_capacities_irena`. However, even if you know exact capacities of existing wind and solar powerplants, currently you cannot set them as `p_nom`. With the proposed PR, it will automatically calculate the aggregated `p_nom` for each bus from `custom_powerplants.csv` and set them accordingly.

## Checklist

- [ ] I consent to the release of this PR's code under the AGPLv3 license and non-code contributions under CC0-1.0 and CC-BY-4.0.
- [ ] I tested my contribution locally and it seems to work fine.
- [ ] Code and workflow changes are sufficiently documented.
- [ ] Newly introduced dependencies are added to `envs/environment.yaml` and `doc/requirements.txt`.
- [ ] Changes in configuration options are added in all of `config.default.yaml` and `config.tutorial.yaml`.
- [ ] Add a test config or line additions to `test/` (note tests are changing the config.tutorial.yaml)
- [ ] Changes in configuration options are also documented in `doc/configtables/*.csv` and line references are adjusted in `doc/configuration.rst` and `doc/tutorial.rst`.
- [ ] A note for the release notes `doc/release_notes.rst` is amended in the format of previous release notes, including reference to the requested PR.

### CODE REVIEW

**Comment by pz-max** [comment]
@yerbol-akhmetov can you share your data somewhere such that we can locally test the contribution? DONE HERE
[custom_powerplants.csv](https://github.com/pypsa-meets-earth/pypsa-earth/files/11578032/custom_powerplants.csv)

File: scripts/add_electricity.py [hunk 1]
```diff
@@ -338,6 +348,24 @@ def attach_wind_and_solar(
                 p_max_pu=ds["profile"].transpose("time", "bus").to_pandas(),
             )
 
+        if suptech == "offwind":
+            continue
+            # df.carrier.mask(df.technology=="Offshore", "offwind-ac",inplace=True)
+
+        df.carrier.mask(df.technology == "Onshore", "onwind", inplace=True)
```
Comment (pz-max) [inline scripts/add_electricity.py]: @yerbol-akhmetov  Is this line just necessary because the custom_powerplant data is not stored correctly / or the docs are described such giving false instructions on storing the data?

File: scripts/add_electricity.py [hunk 2]
```diff
@@ -338,6 +348,24 @@ def attach_wind_and_solar(
                 p_max_pu=ds["profile"].transpose("time", "bus").to_pandas(),
             )
 
+        if suptech == "offwind":
+            continue
+            # df.carrier.mask(df.technology=="Offshore", "offwind-ac",inplace=True)
```
Comment (pz-max) [inline scripts/add_electricity.py]: Why was this commented out?

File: scripts/add_electricity.py [hunk 3]
```diff
@@ -288,11 +289,20 @@ def update_transmission_costs(n, costs, length_factor=1.0, simple_hvdc_costs=Fal
...
+    input_profiles,
+    technologies,
+    extendable_carriers,
+    line_length_factor=1,
 ):
     # TODO: rename tech -> carrier, technologies -> carriers
     _add_missing_carriers_from_costs(n, costs, technologies)
 
+    df = ppl.copy()
+    df.rename(columns={"country": "Country"}, inplace=True)
```
Comment (pz-max) [inline scripts/add_electricity.py]: Why was this necessary?

File: scripts/add_electricity.py [hunk 4]
```diff
@@ -338,6 +348,24 @@ def attach_wind_and_solar(
...
+        gens = n.generators[lambda df: df.carrier == tech]
+        buses = n.buses.loc[gens.bus.unique()]
+        gens_per_bus = gens.groupby("bus").p_nom.count()
+
+        if not gens.empty and not df.query("carrier == @tech").empty:
+            caps = map_country_bus(df.query("carrier == @tech"), buses)
+            caps = caps.groupby(["bus"]).p_nom.sum()
+            caps = caps / gens_per_bus.reindex(caps.index, fill_value=1)
+
+            n.generators.p_nom.update(gens.bus.map(caps).dropna())
```
Comment (pz-max) [inline scripts/add_electricity.py]: why updating the capacities and not add already above at `n.madd()`?

**Review by pz-max — COMMENTED** [review]
Just a few comments. I will test the code locally tomorrow morning for a proper review and code suggestions.

File: scripts/add_electricity.py (hunk 1 above)
Comment (yerbol-akhmetov) [reply scripts/add_electricity.py]: @pz-max, I have added this lines because by default `onwind` is stored as `wind` based on PPM standards. Please, see the figure below: 
image

File: scripts/add_electricity.py (hunk 2 above)
Comment (yerbol-akhmetov) [reply scripts/add_electricity.py]: Hi, @pz-max. I have commented it out to be consistent with previous lines, where offwind is comemnted. Please, see figure below. This portion also need to be modified in the future, because offwind can be AC or DC, but I have not seen in PPM csvs the difference. Not stated if ac or dc.
image
image

File: scripts/add_electricity.py (hunk 3 above)
Comment (yerbol-akhmetov) [reply scripts/add_electricity.py]: The module map_country_bus by default searches "Country" column not "country". Edited.
image

Comment (yerbol-akhmetov) [reply scripts/add_electricity.py]: Here, is the OPSD function: 
image

Comment (yerbol-akhmetov) [reply scripts/add_electricity.py]: I have added this line, because without it was raising error. Maybe it is related to how powerplants.csv is formed. You can try without this line.

File: scripts/add_electricity.py (hunk 4 above)
Comment (yerbol-akhmetov) [reply scripts/add_electricity.py]: @pz-max, we can do that. For that we need to move all added code after condition with hydro: 
image

File: scripts/add_electricity.py (hunk 2 above)
Comment (pz-max) [reply scripts/add_electricity.py]: @yerbol-akhmetov the commented parts for offwind was a hack for updating capital costs.
Here we want to add the capacities of all wind and solar related technologies. 
We need to keep this line/ not skipping the addition of powerplants for offwind.

My suggestion is to do something like:

From:
```
        if suptech == "offwind":
            continue
            # df.carrier.mask(df.technology=="Offshore", "offwind-ac",inplace=True)
```

To:
```
      if tech == "offwind-ac": 
          # add all offwind wind power plants by default as offwind-ac
          df.carrier.mask(df.technology=="Offshore", "offwind-ac",inplace=True)
```

File: scripts/add_electricity.py (hunk 4 above)
Comment (pz-max) [reply scripts/add_electricity.py]: Let's do that :+1: 
Can you make these changes here?

File: scripts/add_electricity.py (hunk 3 above)
Comment (pz-max) [reply scripts/add_electricity.py]: @yerbol-akhmetov let's change the two lines to:
`df = ppl.rename(columns={"country": "Country"})`

File: scripts/add_electricity.py [hunk 5]
```diff
@@ -338,6 +348,24 @@ def attach_wind_and_solar(
                 p_max_pu=ds["profile"].transpose("time", "bus").to_pandas(),
             )
 
+        if suptech == "offwind":
+            continue
+            # df.carrier.mask(df.technology=="Offshore", "offwind-ac",inplace=True)
+
+        df.carrier.mask(df.technology == "Onshore", "onwind", inplace=True)
+
+        gens = n.generators[lambda df: df.carrier == tech]
```
Comment (pz-max) [inline scripts/add_electricity.py]: Instead of the lambda function could you try this:
`n.generators[df.carrier == tech]`

**Review by pz-max — CHANGES_REQUESTED** [review]
Summary of main changes:
- Let's not update the network but read `p_nom` directly into the `n.madd()` if capacities are available
- Let's also read the capacities for offshore technologies and not skipping them. By default, we will read the capacities as `offshore-ac` carrier
- Let's avoid lambda when not necessary

Note. This review also includes feedback on the above comments.

File: scripts/add_electricity.py (hunk 5 above)
Comment (yerbol-akhmetov) [reply scripts/add_electricity.py]: @pz-max, if we put the code in front to make `madd()` instead of `update`, it need to be rewritten majorly. Because currently if searches for already added RES gens in `n.generators ` and updates their p_nom. I will do that.

**Review by pz-max — APPROVED** [review]
